### PR TITLE
fix: pass dev option instead of just include-dev

### PR DESCRIPTION
**Pull request by orsagie**
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
* Basic typescriptyfying of the code.
* Added errors with codes
* Added the codeowners file
* Fixed to pass on the `dev` option in addition to the `include-dev` until we determine that `include-dev` is not used.
* Remove node v4 and add node v10

### CODE REVIEW

File: test/repositories-config-packages.test.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import * as path from 'path';
```
Comment (orsagie) [inline test/repositories-config-packages.test.ts]: Expecting this test to succeed in travis, but fail in appveyor.

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# This is a comment.
+# Each line is a file pattern followed by one or more owners.
+
+# These owners will be the default owners for everything in
+# the repo. Unless a later match takes precedence,
+#  @boost will be requested for
```
Comment (lili2311) [inline .github/CODEOWNERS]: `@boost` => `@snyk/boost`

Comment (orsagie) [reply .github/CODEOWNERS]: changed

File: tslint.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
+    "extends": "tslint:recommended",
+    "rules": {
+        "max-line-length": {
```
Comment (lili2311) [inline tslint.json]: we should use `eslint`, tslint is eol

File: test/stubs/repositories-config/packages/NUnit.2.6.3/license.txt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+Copyright � 2002-2013 Charlie Poole
+Copyright � 2002-2004 James W. Newkirk, Michael C. Two, Alexei A. Vorontsov
+Copyright � 2000-2002 Philip A. Craig
+
+This software is provided 'as-is', without any express or implied warranty. In no event will the authors be held liable for any damages arising from the use of this software.
+
+Permission is granted to anyone to use this software for any purpose, including commercial applications, and to alter it and redistribute it freely, subject to the following restrictions:
```
Comment (lili2311) [inline test/stubs/repositories-config/packages/NUnit.2.6.3/license.txt]: is this needed?

File: test/stubs/repositories-config/expected.json [hunk 5]
```diff
@@ -0,0 +1,24 @@
+{
+  "package": {
+    "dependencies": {
+      "Moq": {
+        "dependencies": {},
```
Comment (lili2311) [inline test/stubs/repositories-config/expected.json]: no transitives?

File: test/parse-dotnet-cli-p-g.test.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+const expectedTree = JSON.parse(fs.readFileSync('./test/stubs/dotnet_p_g/expected.json', 'utf-8'));
+
+test('parse dotnet-cli project and traverse packages', (t) => {
+  plugin.inspect(
+    projectPath,
+    manifestFile,
+    {
+      packagesFolder: path.resolve(projectPath, '_packages'),
+    })
+    .then((result) => {
```
Comment (lili2311) [inline test/parse-dotnet-cli-p-g.test.ts]: Could be a quick `async/await`?

File: test/stubs/repositories-config/packages/NUnit.2.6.3/license.txt (hunk 4 above)
Comment (orsagie) [reply test/stubs/repositories-config/packages/NUnit.2.6.3/license.txt]: no, removing

File: tslint.json (hunk 3 above)
Comment (orsagie) [reply tslint.json]: 👍

File: test/stubs/repositories-config/expected.json (hunk 5 above)
Comment (orsagie) [reply test/stubs/repositories-config/expected.json]: not for the example they provided. I can create one with.

File: lib/nuget-parser/index.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+export async function buildDepTreeFromFiles(
+  root,
+  targetFile,
+  packagesFolderPath,
+  manifestType,
+  useProjectNameFromAssetsFile) {
+  const fileContentPath = path.resolve(root || '.', targetFile || '.');
+  let fileContent;
+  try {
+    debug(`Parsing content of ${fileContentPath}`);
```
Comment (orsagie) [inline lib/nuget-parser/index.ts]: Added this comment to help asses what file we are parsing as a target file.

File: lib/nuget-parser/dotnet-framework-parser.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+export async function parse(tree, manifest, targetFramework, packagesFolder) {
+  if (!targetFramework) {
+    throw new Error('No valid Dotnet target framework found');
+  }
+
+  const flattenedPackageList = scanInstalled(manifest, packagesFolder);
+  return fetchNugetInformationFromPackages(flattenedPackageList, targetFramework)
+    .then(processNugetInformation)
```
Comment (miiila) [inline lib/nuget-parser/dotnet-framework-parser.ts]: Refactor to `async/await`?

File: lib/nuget-parser/dotnet-framework-parser.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+              nuspecResolutions[transitiveDependency.name].children) || [];
+          buildTree(
+            transitiveDependency,
+            transitiveChildren,
+            repository);
+          node.dependencies[transitiveDependency.name] = transitiveDependency;
+        });
+      }
+
+      const nugtKeys = Object.keys(nuspecResolutions);
```
Comment (miiila) [inline lib/nuget-parser/dotnet-framework-parser.ts]: typo? I'd heavily prefer `nugetKeys`

File: lib/nuget-parser/nuspec-parser.ts [hunk 10]
```diff
@@ -1,43 +1,42 @@
...
 
-          let ownDeps = [];
+          let ownDeps: any = [];
           // We are only going to check the first targetFramework we encounter
           // in the future we may want to support multiple, but only once
           // we have dependency version conflict resolution implemented
-          // _(targetFrameworks).forEach(function (targetFramework) {
-          _(result.package.metadata).forEach(function (metadata) {
-            _(metadata.dependencies).forEach(function (rawDependency) {
+          // _(targetFrameworks).forEach((targetFramework) => {
```
Comment (miiila) [inline lib/nuget-parser/nuspec-parser.ts]: Please delete this code if it won't be used

File: lib/nuget-parser/nuspec-parser.ts [hunk 11]
```diff
@@ -82,42 +80,43 @@ function parseNuspec(dep, targetFramework, sep) {
...
     });
 }
 
 function extractDepsFromRaw(rawDependencies) {
-  const deps = [];
-  _.forEach(rawDependencies, function (dependency) {
-    if (dependency && dependency.$) {
-      const dep = new Dependency(dependency.$.id, dependency.$.version);
-      deps.push(dep);
+  const deps: dependency.Dependency[] = [];
```
Comment (miiila) [inline lib/nuget-parser/nuspec-parser.ts]: can we import `{Dependency}` to get rid of `dependency.Dependency`?

File: lib/nuget-parser/project-json-parser.ts [hunk 12]
```diff
@@ -22,17 +22,17 @@ function scanForDependencies(obj, deps) {
...
     }
-  });
+  }
   return deps;
 }
 
 function parseJsonManifest(fileContent) {
   const rawContent = JSON.parse(fileContent);
-  const result = {};
+  const result: any = {};
```
Comment (miiila) [inline lib/nuget-parser/project-json-parser.ts]: Another missing interface we should know or create

File: package.json [hunk 13]
```diff
@@ -17,6 +21,13 @@
   ],
   "author": "snyk.io",
   "license": "Apache-2.0",
+  "engines": {
+    "node": ">=4"
```
Comment (miiila) [inline package.json]: I believe we've deprecated node 4, so it should be `>=6`

File: tsconfig.json [hunk 14]
```diff
@@ -0,0 +1,19 @@
+{
+  "compilerOptions": {
+    "outDir": "./dist",
+    "pretty": true,
+    "target": "es2015",
+    "lib": [
+      "es2015",
+      "es2016.array.include",
```
Comment (miiila) [inline tsconfig.json]: Once moved to node6, I hope this is not necessary anymore :-)

File: lib/nuget-parser/csproj-parser.ts [hunk 15]
```diff
@@ -1,22 +1,24 @@
...
   const csprojPath = findFile(rootDir, /.*\.csproj$/);
   if (csprojPath) {
     debug('Checking .net framework version in .csproj file ' + csprojPath);
 
     const csprojContents = fs.readFileSync(csprojPath);
 
-    let frameworks = [];
-    parseXML(csprojContents, function (err, parsedCsprojContents) {
+    let frameworks: any = [];
+    parseXML.parseString(csprojContents, (err, parsedCsprojContents) => {
```
[the remark this answers is not in the document]
Comment (orsagie) [reply lib/nuget-parser/csproj-parser.ts]: This should not have been an async method.

In addition xml2js does not seem to have a non-callback way to call it.

File: lib/nuget-parser/csproj-parser.ts [hunk 16]
```diff
@@ -1,22 +1,24 @@
...
   debug('Looking for your .csproj file in ' + rootDir);
   const csprojPath = findFile(rootDir, /.*\.csproj$/);
   if (csprojPath) {
     debug('Checking .net framework version in .csproj file ' + csprojPath);
 
     const csprojContents = fs.readFileSync(csprojPath);
 
-    let frameworks = [];
-    parseXML(csprojContents, function (err, parsedCsprojContents) {
+    let frameworks: any = [];
```
[the remark this answers is not in the document]
Comment (orsagie) [reply lib/nuget-parser/csproj-parser.ts]: changing

File: lib/nuget-parser/csproj-parser.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (miiila) [reply lib/nuget-parser/csproj-parser.ts]: Got it, I meant something like:
```
function parseXMLstring(fileContent: string) {
  return new Promise((resolve, reject) => {
    parseXML.parseString(fileContent, (err, result) => {
      return err ? reject(err) : resolve(result);
    })
  })
}
```

But it's not critical.

[the remark this answers is not in the document]
Comment (orsagie) [reply lib/nuget-parser/csproj-parser.ts]: gotcha, can change

File: lib/nuget-parser/dotnet-framework-parser.ts (hunk 8 above)
Comment (orsagie) [reply lib/nuget-parser/dotnet-framework-parser.ts]: changed

File: lib/nuget-parser/dotnet-framework-parser.ts (hunk 9 above)
Comment (orsagie) [reply lib/nuget-parser/dotnet-framework-parser.ts]: changing

File: lib/nuget-parser/nuspec-parser.ts (hunk 11 above)
Comment (orsagie) [reply lib/nuget-parser/nuspec-parser.ts]: yes

File: lib/nuget-parser/nuspec-parser.ts (hunk 10 above)
Comment (orsagie) [reply lib/nuget-parser/nuspec-parser.ts]: removed

File: lib/index.ts [hunk 17]
```diff
@@ -51,8 +51,8 @@ module.exports = {
         root,
         targetFile,
         path.join(path.dirname(targetFile), 'paket.lock'),
-        options['include-dev'],
-        options.strict
+        options['include-dev'] || options.dev, // TODO: remove include-dev when no longer used.
```
Comment (orsagie) [inline lib/index.ts]: @miiila here is the fix

Comment (miiila) [reply lib/index.ts]: @orsagie I found it! 😂

**Pull request merged**